### PR TITLE
Simplify/optimize Int128 comparison operators

### PR DESCRIPTION
**Pull request by lilinus**
We can remove a some branches in comparison operators for `Int128` by using signed comparison for upper 64-bit values.
Maybe this approach was already considered but not chosen, in that case I will close the PR.

Also added some test cases for some better coverage on said operators.

It was hard to benchmark any performance gain because (getting ZeroMeasurement) so I had to run the operators inside a loop.
Link to benchmark source code [here](https://github.com/lilinus/performance/blob/1b0e512f1bd447c167a542042c8d36a0151e6661/src/benchmarks/micro/libraries/System.Runtime/Perf.TestPr.cs).
Below are benchmarks for `>` operator (other operators gives similar results):
| Method         | Job        | Toolchain | val1                                 | val2 | Mean     | Error    | StdDev   | Median   | Min      | Max      | Ratio | Allocated | Alloc Ratio |
|--------------- |----------- |---------- |------------------------------------- |----- |---------:|---------:|---------:|---------:|---------:|---------:|------:|----------:|------------:|
| op_GreaterThan | Job-QLZCZD | main      | -170141183460(...)3715884105728 [40] | -1   | 28.16 ns | 0.092 ns | 0.086 ns | 28.15 ns | 28.04 ns | 28.32 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | -170141183460(...)3715884105728 [40] | -1   | 27.86 ns | 0.046 ns | 0.043 ns | 27.87 ns | 27.81 ns | 27.95 ns |  0.99 |         - |          NA |
| op_GreaterThan | Job-QLZCZD | main      | -170141183460(...)3715884105728 [40] | 1    | 27.89 ns | 0.061 ns | 0.051 ns | 27.91 ns | 27.83 ns | 27.97 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | -170141183460(...)3715884105728 [40] | 1    | 28.02 ns | 0.072 ns | 0.067 ns | 28.02 ns | 27.92 ns | 28.13 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-QLZCZD | main      | -1                                   | -1   | 28.90 ns | 0.088 ns | 0.073 ns | 28.94 ns | 28.71 ns | 28.97 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | -1                                   | -1   | 27.85 ns | 0.047 ns | 0.041 ns | 27.86 ns | 27.80 ns | 27.93 ns |  0.96 |         - |          NA |
| op_GreaterThan | Job-QLZCZD | main      | -1                                   | 1    | 27.86 ns | 0.064 ns | 0.054 ns | 27.85 ns | 27.73 ns | 27.93 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | -1                                   | 1    | 27.86 ns | 0.068 ns | 0.064 ns | 27.84 ns | 27.75 ns | 28.00 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-QLZCZD | main      | 0                                    | -1   | 27.85 ns | 0.054 ns | 0.051 ns | 27.87 ns | 27.79 ns | 27.92 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | 0                                    | -1   | 28.01 ns | 0.096 ns | 0.080 ns | 28.00 ns | 27.86 ns | 28.11 ns |  1.01 |         - |          NA |
| op_GreaterThan | Job-QLZCZD | main      | 0                                    | 1    | 28.02 ns | 0.083 ns | 0.078 ns | 27.98 ns | 27.88 ns | 28.14 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | 0                                    | 1    | 27.84 ns | 0.055 ns | 0.049 ns | 27.84 ns | 27.79 ns | 27.94 ns |  0.99 |         - |          NA |
| op_GreaterThan | Job-QLZCZD | main      | 1                                    | -1   | 27.81 ns | 0.087 ns | 0.081 ns | 27.82 ns | 27.68 ns | 27.98 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | 1                                    | -1   | 28.03 ns | 0.067 ns | 0.063 ns | 28.04 ns | 27.88 ns | 28.12 ns |  1.01 |         - |          NA |
| op_GreaterThan | Job-QLZCZD | main      | 1                                    | 1    | 27.83 ns | 0.087 ns | 0.078 ns | 27.84 ns | 27.64 ns | 27.97 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | 1                                    | 1    | 28.00 ns | 0.104 ns | 0.086 ns | 28.02 ns | 27.82 ns | 28.10 ns |  1.01 |         - |          NA |
| op_GreaterThan | Job-QLZCZD | main      | 1701411834604(...)3715884105727 [39] | -1   | 28.92 ns | 0.066 ns | 0.059 ns | 28.92 ns | 28.76 ns | 28.99 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | 1701411834604(...)3715884105727 [39] | -1   | 27.58 ns | 0.054 ns | 0.051 ns | 27.59 ns | 27.50 ns | 27.66 ns |  0.95 |         - |          NA |
| op_GreaterThan | Job-QLZCZD | main      | 1701411834604(...)3715884105727 [39] | 1    | 27.79 ns | 0.082 ns | 0.077 ns | 27.81 ns | 27.66 ns | 27.92 ns |  1.00 |         - |          NA |
| op_GreaterThan | Job-PVMDKE | pr        | 1701411834604(...)3715884105727 [39] | 1    | 27.64 ns | 0.061 ns | 0.054 ns | 27.63 ns | 27.54 ns | 27.74 ns |  0.99 |         - |          NA |

In some cases it seems we get ~5% perf improvements.
Code (and codegen) is more compact. [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8YUCAAcAymIBuvMDFwBuGsQDMTUgwDCDAN40GxhgG0pGKBzAYAMtgCeEDhgAU9py4DSvLgBMWGQBHDhgeXmwAGwBKAF0jEw0GWGw/CC5IhwZcS2sMBiEE40NqEzKGXwwxLiiGSC4cip4GKV4ALxgGAF4GciRVahoAYl4AMwYAISEAcQBRADkAESEAQXmi8okoXllsKuSYVPTMhg5I9IBzBgB9DgkJMQHyky2dvc6UtIyss8ub84A7o9hjBIrgYBsyq9dvtPscfucuFdroDgaVngxoe8DkdvqdEci7g8oAMhmE/GNIYlNMwkAxgBAIJEGNMUlUoAAVAAW2C41wACmxXEIGJEYKMMGgCsleBcuRholTikrysQAOwMVyuAnRMUSli3e5iBgAPk1Ou2cowBqJYkV6IxGMIhE12vF1sNxO6PUt8ptRqg0QYADJg+b3QbUVBTTKrZGIEDA9EnhiAL4qpK0+mM5msw7s7m864AWWwvmFovdUpFvoVKpKjrKY01Qlw8xgFz2OxgbolQa6PVb7c7fFkPdr0XtjZMDenqo1vY9tujZtr/uJU7njudrr1S4D3tjfs9dpDYcX8cTMbXKITduTKrK6YdjdB4MfM4/z3VBTbHa7Y6uBOKbTs+jpgRimZ9NmTK/sOAE9iKuyRKE/ZmtqlzRMhoTrsaAA8DB0CBDDPqmQA==).

### CODE REVIEW

**Comment by tannergooding** [comment]
Your benchmark code is invalid and not testing what you think it's testing.

Since the result of `val1 > val2` is unused, the JIT is able to optimize it out resulting in you testing an empty loop: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgNIAO1wYGAybIFioogRGkMAvA3IAGXQG4uPY93pMUDCLwD6AcViiYUACoALbJIAUASQXlSAA4GEQAbcjQGPwwA4LDSAEpTDmSeBgAzaAYveUUtBkM5BgAeOWVVdVwDOQBqGqTqNKaeG3yw8gYAPhDsUNIjRp4AXxohoA==

You can somewhat fix that by doing something like this, where you combine all the comparisons together: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgNIAO1wYGAybIFioogRGkMAvA3IAGXQG4uPY93oNgECABsGEXgH0A4rFEwoAFQAW2SQAoASQVyUgAOBhFrcjQGIIwQ8MjSAEpTDjSeCytbWFwAV2tZbQAzbGtcGCNqTMyMnmLoBj95RS0GQzkGAB45ZVV1XAM5AGph1OqaydyC2UJtSPIGAD4IstIqye46swB2BmnCjZ4AXxpjoA==

However, that itself is also a "point in time" consideration as some runtimes or future versions of RyuJIT may introspect that `val1 > val2` is invariant and therefore it could be rewritten to simply `return val1 > val2;`

**Review by tannergooding — APPROVED** [review]
Code is purely simpler and is still correct

That being said, it's worth noting that this simplification doesn't buy much and the real optimization requires some JIT work so we can simply emit a `cmp; sbb; setcc` sequence instad.

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs [hunk 1]
```diff
@@ -1043,57 +1043,29 @@ public static Int128 Log2(Int128 value)
...
-            if (IsNegative(left) == IsNegative(right))
-            {
-                return (left._upper < right._upper)
-                    || ((left._upper == right._upper) && (left._lower < right._lower));
-            }
-            else
-            {
-                return IsNegative(left);
-            }
+            return ((long)left._upper < (long)right._upper)
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Int128.cs:1046]: A comment (in the code) explaining why this is valid would be beneficial.

Actually brute forcing this for some `struct S16 { byte _lower; byte _upper; }` is pretty trivial to do, but so is describing why this works out given the two's complement representation. -- Notably this comes down to the fact that all negative values have the sign bit set, so if `lhs.upper < rhs.upper` in signed form, then either lhs/rhs have different signs, and the single comparison was sufficient or they have the same sign and the same was true.

------

There may be other optimizations available here as well... In particular, comparisons `x cmp y` in general are done as if `x - y` and then setting CPU flags, without actually producing the result. That general premise is why the "most optimal" pattern is actually `cmp lower; sbb upper; setcc`

Comment (lilinus) [reply src/libraries/System.Private.CoreLib/src/System/Int128.cs:1053]: I tried my best documenting it with comments.

**Pull request merged**